cpp/src/arrow/util/spaced.h

```diff
@@ -0,0 +1,266 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "arrow/util/bit_util.h"
+
+#ifdef ARROW_HAVE_AVX512
+#include <immintrin.h>
+#endif  // ARROW_HAVE_AVX512
+
+namespace arrow {
+namespace util {
+namespace internal {
+
+template <typename T>
+int PutSpacedScalar(const T* values, int num_values, const uint8_t* valid_bits,
```

Comment:
FWIW, Daniel Lemire has a SIMD library for this https://lemire.me/blog/2017/04/25/quickly-pruning-elements-in-simd-vectors-using-the-simdprune-library/

Reply:
might be worth a quick discussion on whether we should vendor this (or maybe just vendor the lookup tables).

Reply:
> might be worth a quick discussion on whether we should vendor this (or maybe just vendor the lookup tables).

The tables for epi8 is really large, it use a 1MB space. The small size of int32/int64 path seems more valuable.

Reply:
yeah, I think there is a thin version for epi8 but not sure if we need anything but int32/int64 either way before doing something we should discuss on the mailing list.

Reply:
SSE epi8/epi32/epi64 added.